File: server/app/run_code_scripts/run_code.js
const { exec } = require('child_process');
const fs = require('fs');
const path = require('path');

const code = process.argv[2];
const input = process.argv[3];

const codePath = path.join(__dirname, 'temp_code.js');
const inputPath = path.join(__dirname, 'temp_input.txt');

fs.writeFileSync(codePath, code);
fs.writeFileSync(inputPath, input);

exec(`node ${codePath} < ${inputPath}`, (error, stdout, stderr) => {
    const exitCode = error ? error.code : 0;
    const response = {
        output: stdout,
        error: stderr,
        exit_code: exitCode
    };
    console.log(JSON.stringify(response));

    fs.unlinkSync(codePath);
    fs.unlinkSync(inputPath);
});

File: server/app/run_code_scripts/run_code.py
import subprocess
import sys
import json
import tempfile

code = sys.argv[1]
input_data = sys.argv[2]

with tempfile.NamedTemporaryFile(delete=False, suffix=".py") as tmp_code:
    tmp_code.write(code.encode('utf-8'))
    tmp_code_path = tmp_code.name

with tempfile.NamedTemporaryFile(delete=False, suffix=".txt") as tmp_input:
    tmp_input.write(input_data.encode('utf-8'))
    tmp_input_path = tmp_input.name

result = subprocess.run(['python3', tmp_code_path], stdin=open(tmp_input_path), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
output = result.stdout.decode('utf-8')
error = result.stderr.decode('utf-8')
exit_code = result.returncode

response = {
    'output': output,
    'error': error,
    'exit_code': exit_code
}

print(json.dumps(response))

os.remove(tmp_code_path)
os.remove(tmp_input_path)
